server/aesdsocket.c

```diff
@@ -0,0 +1,326 @@
+	saddr.sin_port = htons(MYPORT);
+
+	int ret = bind(sockfd, (struct sockaddr *) &saddr, sizeof(struct sockaddr_in));
+	if (ret == -1)
+	{
+		syslog(LOG_ERR, "socket binding failed\n");
+		closeall();
+		exit(EXIT_FAILURE);
+	}
+	syslog(LOG_INFO, "bind successful\n");
+
+	// start deamon
+	if (deamon) 
+	{
+        pid_t pid = fork();
+        if (pid == -1) 
+		{
+            syslog(LOG_ERR, "failed to fork");
+			closeall();
+            exit(EXIT_FAILURE);
+        }
+        else if (pid > 0)
+		{
+            exit(EXIT_SUCCESS);
+        }
+
+		syslog(LOG_INFO, "fork successful\n");
+	
+        pid_t sid = setsid();		
+		if (sid == -1) 
+		{
+			syslog(LOG_ERR, "failed to setsid");
+			closeall();
+			exit(EXIT_FAILURE);
+		}
+		
+		syslog(LOG_INFO, "SID: %d\n", sid);
+
+		if (chdir("/") == -1) 
+		{
+			syslog(LOG_ERR, "failed to change dir");
+			closeall();
+			exit(EXIT_FAILURE);
+		}
+		syslog(LOG_INFO, "chdir successful\n");
+
+		//open("/dev/null", O_RDWR);
+		//dup(STDIN_FILENO);
+		//dup(STDOUT_FILENO);
+		//dup(STDERR_FILENO);
+
+		close(STDIN_FILENO);
+		close(STDOUT_FILENO);
+		close(STDERR_FILENO);
+    }
+	syslog(LOG_INFO, "daemon created\n");
+	
+	ret = listen(sockfd, 10);
+	if (ret == -1)
+	{
+		syslog(LOG_ERR, "socket listening failed\n");
+		closeall();
+		exit(EXIT_FAILURE);
+	}
+	syslog(LOG_INFO, "listening\n");
+	
+	socklen_t len = sizeof(struct sockaddr);
+
+	fd = open(filepath, O_CREAT | O_RDWR | O_APPEND | O_TRUNC, 0764);
+	if (fd == -1)	
+	{//if error
+		syslog(LOG_ERR, "can't open or create file '%s'\n", filepath);
+		closeall();
+		exit(EXIT_FAILURE);
+	}
+	syslog(LOG_DEBUG, "opened or created file '%s' successfully\n", filepath);
+	
+	while(1)
+	{
+		acceptedfd = accept(sockfd, (struct sockaddr *) &saddr, &len);
+		if (acceptedfd == -1)
+		{
+			syslog(LOG_ERR, "socket accepting failed\n");
+			closeall();
+			exit(EXIT_FAILURE);
+		}
+
+		syslog(LOG_DEBUG, "Accepted connection from '%s'\n", inet_ntoa((struct in_addr)saddr.sin_addr));
+	
+
+		//Signals from set are added to invoking process's signal mask
+        if (sigprocmask(SIG_BLOCK, &mask, NULL)) 
+		{
+            syslog(LOG_ERR, "sigprocmask");
+			closeall();
+            exit(EXIT_FAILURE);
+        }
+
+		do
+		{
+			memset(buffer, 0, sizeof(buffer));
+			
+			nbytes = recv(acceptedfd, buffer, sizeof(buffer), 0);
+
+			if (nbytes)
+			{
+				// syslog(LOG_DEBUG, "nbytes - %d\n", nbytes);
+				// syslog(LOG_DEBUG, "'%s'\n", buffer);
+				if(nbytes > strlen(buffer))
+					nr = write(fd, buffer, strlen(buffer));
+				else
+					nr = write(fd, buffer, nbytes);
+					
+				if (nr == -1)	
+				{//if error
+					syslog(LOG_ERR, "can't write received string in file '%s'\n", filepath);
+					break;	
+				}
+				total_bytes += nr;
+			}
+			else
+				break;
+
+		}while(strchr(buffer, '\n') == NULL);
+		
+				
+		nbytes=0;
+		lseek(fd, 0, SEEK_SET);
+		
+		sendbuffer = (char *)malloc(sizeof(char) * (total_bytes+location));
```

Comment:
@MuktaDarekar what assumption does this make about the total size of packets sent relative to the size of the heap?

Reply:
modified